### PR TITLE
feat(swift): add vulns for cocoapods

### LINKED ISSUE #4896: Add support for Swift packages vulnerabilities
## Description
Add support for detecting vulnerabilities in Swift packages. It should support the following package managers:
- [CocoaPods](https://cocoapods.org/)
- [Swift Package Manager](https://www.swift.org/package-manager/)

### Discussed in https://github.com/aquasecurity/trivy/discussions/4895

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
Add vulnerability detection for `Cocoapods`.
Update `purl` - use `subpath`. 
```
➜ trivy fs ./Podfile.lock 
2023-08-25T12:17:43.021+0600    INFO    Vulnerability scanning is enabled
2023-08-25T12:17:43.022+0600    INFO    Secret scanning is enabled
2023-08-25T12:17:43.022+0600    INFO    If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-08-25T12:17:43.022+0600    INFO    Please see also https://aquasecurity.github.io/trivy/dev/docs/scanner/secret/#recommendation for faster secret detection
2023-08-25T12:17:43.027+0600    INFO    Number of language-specific files: 1
2023-08-25T12:17:43.027+0600    INFO    Detecting cocoapods vulnerabilities...

Podfile.lock (cocoapods)

Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 1, HIGH: 0, CRITICAL: 0)

┌────────────────────┬───────────────┬──────────┬────────┬───────────────────┬────────────────────────┬────────────────────────────────────────────────────────┐
│      Library       │ Vulnerability │ Severity │ Status │ Installed Version │     Fixed Version      │                         Title                          │
├────────────────────┼───────────────┼──────────┼────────┼───────────────────┼────────────────────────┼────────────────────────────────────────────────────────┤
│ _NIODataStructures │ CVE-2022-3215 │ MEDIUM   │ fixed  │ 2.41.0            │ 2.29.1, 2.39.1, 2.42.0 │ SwiftNIO vulnerable to Improper Neutralization of CRLF │
│                    │               │          │        │                   │                        │ Sequences in HTTP Headers ('HTTP...                    │
│                    │               │          │        │                   │                        │ https://avd.aquasec.com/nvd/cve-2022-3215              │
└────────────────────┴───────────────┴──────────┴────────┴───────────────────┴────────────────────────┴────────────────────────────────────────────────────────┘

```

## Related issues
- Close #4896

## Related PRs
- [x] aquasecurity/trivy-db/pull/344
- [x] aquasecurity/go-dep-parser/pull/252

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/sbom/cyclonedx/unmarshal_test.go [hunk 1]
```diff
@@ -313,6 +313,50 @@ func TestUnmarshaler_Unmarshal(t *testing.T) {
 				},
 			},
 		},
+		{
+			name:      "happy path for cocoapods with purl subpath",
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/sbom/cyclonedx/unmarshal_test.go:317]: We don't test [Package()](https://github.com/aquasecurity/trivy/blob/9345a98ed1125d022e057d2c4066df0e17662cf1/pkg/purl/purl.go#L41) function in `purl` package.
I only found 1 my mistake with `subpath` when testing SBOM manually.
That is why i added tests in SBOM.

[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/sbom/cyclonedx/unmarshal_test.go:317]: Oh, we should add `TestPackage` then.

[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/sbom/cyclonedx/unmarshal_test.go]: Removed added test for SBOM and added `TestPackage`.

File: docs/docs/coverage/language/swift.md [hunk 2]
```diff
@@ -1,10 +1,34 @@
...
+|    Cocoapods    | Podfile.lock     |            ✓            |     Included     |                  ✓                   |    -     |
+
+These may be enabled or disabled depending on the target.
+See [here](./index.md) for the detail.
+
+## swift
+Trivy parses [Package.resolved][package-resolved] file to find dependencies. Don't forger to update (`swift package update` command) this file before scanning.
+
+## cocoapods
+Cocoapods uses package names in `PodFile.lock`, but [GitHub advisory database][ghsa] uses git links.
```
Comment (knqyf263) [inline docs/docs/coverage/language/swift.md:26]: We should mention the limitation that all modules are detected under the same git URL.

File: docs/docs/coverage/language/swift.md [hunk 3]
```diff
@@ -1,10 +1,34 @@
...
+|-----------------|:----:|:-------------:|:-------:|
+| Swift           |  ✓   |       ✓       |    -    |
+| CocoaPods       |  ✓   |       ✓       |    -    |
+
+The following table provides an outline of the features Trivy offers.
+
+| Package manager | File             | Transitive dependencies | Dev dependencies | [Dependency graph][dependency-graph] | Position |
+|:---------------:|------------------|:-----------------------:|:----------------:|:------------------------------------:|:--------:|
+|      Swift      | Package.resolved |            ✓            |     Included     |                  -                   |    ✓     |
+|    Cocoapods    | Podfile.lock     |            ✓            |     Included     |                  ✓                   |    -     |
```
Comment (knqyf263) [inline docs/docs/coverage/language/swift.md:17]:
```suggestion
|    CocoaPods    | Podfile.lock     |            ✓            |     Included     |                  ✓                   |    -     |
```

File: docs/docs/coverage/language/swift.md [hunk 4]
```diff
@@ -1,10 +1,34 @@
...
+
+These may be enabled or disabled depending on the target.
+See [here](./index.md) for the detail.
+
+## swift
```
Comment (knqyf263) [inline docs/docs/coverage/language/swift.md:22]:
```suggestion
## Swift
```

File: docs/docs/coverage/language/swift.md [hunk 5]
```diff
@@ -1,10 +1,34 @@
...
+Trivy parses [Package.resolved][package-resolved] file to find dependencies. Don't forger to update (`swift package update` command) this file before scanning.
+
+## cocoapods
```
Comment (knqyf263) [inline docs/docs/coverage/language/swift.md:25]:
```suggestion
## CocoaPods
```

File: docs/docs/coverage/language/swift.md [hunk 6]
```diff
@@ -1,10 +1,34 @@
...
+Cocoapods uses package names in `PodFile.lock`, but [GitHub advisory database][ghsa] uses git links. 
+We parse [Cocoapods Specs][cocoapods-specs] to match package names and links.
 
-[cocoapods]: https://cocoapods.org/
\ No newline at end of file
+[cocoapods]: https://cocoapods.org/
+[cocoapods-specs]: https://github.com/CocoaPods/Specs
+[ghsa]: https://github.com/advisories?query=type%3Areviewed+ecosystem%3Aswift
+[swift]: https://www.swift.org/package-manager/
+[package-resolved]: https://github.com/apple/swift-package-manager/blob/main/Documentation/Usage.md#resolving-versions-packageresolved-file
```
Comment (knqyf263) [inline docs/docs/coverage/language/swift.md:33]: The link should be permanent. The main branch can be updated.

File: docs/docs/coverage/language/swift.md (hunk 2 above)
Comment (knqyf263) [inline docs/docs/coverage/language/swift.md:26]:
```suggestion
CocoaPods uses package names in `PodFile.lock`, but [GitHub Advisory Database][ghsa] uses git links. 
```

File: pkg/purl/purl.go [hunk 7]
```diff
@@ -57,6 +57,15 @@ func (p *PackageURL) Package() *ftypes.Package {
 		}
 	}
 
+	// Cocoapods purl has no namespace, but has subpath
+	// https://github.com/package-url/purl-spec/blob/a748c36ad415c8aeffe2b8a4a5d8a50d16d6d85f/PURL-TYPES.rst#cocoapods
+	if p.Type == packageurl.TypeCocoapods && p.Subpath != "" {
+		pkg.Name = strings.Join([]string{
+			p.Name,
+			p.Subpath,
```
Comment (knqyf263) [inline pkg/purl/purl.go:65]: Isn't the subpath `#Twitter` in the example of `pkg:cocoapods/ShareKit@2.0#Twitter`?

File: docs/docs/coverage/language/swift.md (hunk 3 above)
Comment (DmitriyLewen) [reply docs/docs/coverage/language/swift.md]: Done

File: docs/docs/coverage/language/swift.md (hunk 4 above)
Comment (DmitriyLewen) [reply docs/docs/coverage/language/swift.md]: Done

File: docs/docs/coverage/language/swift.md (hunk 5 above)
Comment (DmitriyLewen) [reply docs/docs/coverage/language/swift.md]: Done

File: docs/docs/coverage/language/swift.md (hunk 6 above)
Comment (DmitriyLewen) [reply docs/docs/coverage/language/swift.md]: Done

File: docs/docs/coverage/language/swift.md (hunk 2 above)
Comment (DmitriyLewen) [reply docs/docs/coverage/language/swift.md]: Done

File: pkg/purl/purl.go (hunk 7 above)
Comment (DmitriyLewen) [reply pkg/purl/purl.go:65]: I'm not sure I understand what you mean.
`Twitter` is subpath for `pkg:cocoapods/ShareKit@2.0#Twitter`
`NSData+zlib`is subpath for `pkg:cocoapods/GoogleUtilities@7.5.2#NSData+zlib`

Comment (knqyf263) [reply pkg/purl/purl.go:65]: So, why do you join name and subpath with `/`?

Comment (DmitriyLewen) [reply pkg/purl/purl.go:65]: `CocoaPods` uses `<ModuleName>/<SubmoduleName>` format.
e.g for `GoogleUtilities`:
```
➜  cat Podfile.lock 
PODS:
  - GoogleUtilities (7.5.2):
    - GoogleUtilities/AppDelegateSwizzler (= 7.5.2)
    - GoogleUtilities/Environment (= 7.5.2)
    - GoogleUtilities/ISASwizzler (= 7.5.2)
    - GoogleUtilities/Logger (= 7.5.2)
    - GoogleUtilities/MethodSwizzler (= 7.5.2)
    - GoogleUtilities/Network (= 7.5.2)
    - "GoogleUtilities/NSData+zlib (= 7.5.2)"
...
```
I join `module` and `submodule` with `/` to reproduce this format.

Comment (DmitriyLewen) [reply pkg/purl/purl.go:65]: I will add comment for this

Comment (knqyf263) [reply pkg/purl/purl.go:65]: Got it. Thanks.

Comment (DmitriyLewen) [reply pkg/purl/purl.go:67]: Added in 775a6c39af9f14214fdf1c08f322b50c12194760

Comment (knqyf263) [reply pkg/purl/purl.go:67]: Thanks! BTW, is GoogleUtilities a pod name, a module name or a spec name? Is `GoogleUtilities/AppDelegateSwizzler` a subpod, submodule or subspec? "Module might be used in Swift, but not used in CocoaPods. I just want to make sure of the official name.
https://guides.cocoapods.org/syntax/podspec.html

Comment (knqyf263) [reply pkg/purl/purl.go:67]: We'll go with "module" and fix it later if needed.

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**